compiler/src/dmd/dtemplate.d

```diff
@@ -4129,6 +4128,65 @@ MATCH deduceType(RootObject o, Scope* sc, Type tparam, TemplateParameters* param
                                         goto L2;
                                 }
                             }
+                            if (TemplateDeclaration td = s.isTemplateDeclaration())
```

Comment:
This is very hard to review. I suggest taking the code from 4132 .. 4189 and pulling it out into a separate function, which includes documentation for what it does.

Reply:
> This is very hard to review. I suggest taking the code from 4132 .. 4189 and pulling it out into a separate function, which includes documentation for what it does.

I've pulled out part of it. Would you have another look?